### PR TITLE
Update ipsec.widget.php - Count user in "Overview" Tab and improve "Mobile Users" Tab

### PR DESCRIPTION
**Pull request by Monviech**
https://github.com/opnsense/core/issues/6889

This change attempts to solve the problem of users having multiple IP addresses as leases and being counted per leased IP.
- Only "user" are counted now
- Each "user" can have several "lease"
- Each "lease" can have an individual online or offline status
- A user is online when at least one "lease" is "online = true"

The code can be tested with this sample data. It can be placed in line 46 and populates the variable $ipsec_leases = []

```
$ipsec_leases = array(
    array(
        'pool' => 'pool-roadwarrior-ipv4',
        'address' => '172.16.203.1',
        'online' => true,
        'user' => 'user1@vpn1'
    ),
    array(
        'pool' => 'pool-roadwarrior-ipv4',
        'address' => '172.16.203.2',
        'online' => false,
        'user' => 'user2@vpn1'
    ),
    array(
        'pool' => 'pool-roadwarrior-ipv4',
        'address' => '172.16.203.3',
        'online' => true,
        'user' => 'user3@vpn1'
    ),
    array(
        'pool' => 'pool-roadwarrior-ipv4',
        'address' => '172.16.203.4',
        'online' => true,
        'user' => 'user5@vpn1'
    ),
    array(
        'pool' => 'pool-roadwarrior-ipv6',
        'address' => 'fd00:abcd:1234:1::1',
        'online' => false,
        'user' => 'user1@vpn1'
    ),
    array(
        'pool' => 'pool-roadwarrior-ipv6',
        'address' => 'fd00:abcd:1234:1::2',
        'online' => false,
        'user' => 'user1@vpn1'
    ),
    array(
        'pool' => 'pool-roadwarrior-ipv6',
        'address' => 'fd00:abcd:1234:1::3',
        'online' => true,
        'user' => 'user1@vpn1'
    ),
    array(
        'pool' => 'pool-roadwarrior-ipv6',
        'address' => 'fd00:abcd:1234:2::1',
        'online' => false,
        'user' => 'user2@vpn1'
    ),
    array(
        'pool' => 'pool-roadwarrior-ipv6',
        'address' => 'fd00:abcd:1234:2::2',
        'online' => false,
        'user' => 'user3@vpn1'
    ),
    array(
        'pool' => 'pool-roadwarrior-ipv6',
        'address' => 'fd00:abcd:1234:3::1',
        'online' => true,
        'user' => 'user4@vpn1'
    )
);
```
The result will look like this:

grafik

All users that have at least one online lease will be counted towards "Mobile Users". In the above sample data, user2@vpn1 has two offline leases, which means the user counts as not connected. All other users have at least one or multiple online leases, which makes them count as online.

grafik

I think this is the best variant, since it adds new functionality while also retaining all old functionality. :)

### CODE REVIEW

File: src/www/widgets/widgets/ipsec.widget.php [hunk 1]
```diff
@@ -68,7 +68,19 @@
         }
     }
 }
-
+// Initialize variable aggregated_data and loop through the ipsec_leases array to fetch the data for user, address and online status. Used later for div ipsec-mobile.
+$aggregated_data = array();
```
Comment (AdSchellevis) [inline src/www/widgets/widgets/ipsec.widget.php:72]: `$aggregated_data = [];` is a bit shorter, the `array()` construction is used less often in our newer code.

File: src/www/widgets/widgets/ipsec.widget.php [hunk 2]
```diff
@@ -68,7 +68,19 @@
...
+foreach ($ipsec_leases as $lease) {
+    $user = $lease['user'];
+    $address = $lease['address'];
+    $online = $lease['online'];
+    // For each unique user, initialize an empty array
+    if (!isset($aggregated_data[$user])) {
```
Comment (AdSchellevis) [inline src/www/widgets/widgets/ipsec.widget.php:78]: It's shorter to use the `$lease['user']` without storing it first (e.g. `if (!isset($aggregated_data[$lease['user']])) {`), although for `$user` I can understand it if one would find it more readable, the address and online fields are certainly less useful to store first (just adding `$lease` to `$aggregated_data[$user]` might also be an option, the extra user field is not used anyway.

File: src/www/widgets/widgets/ipsec.widget.php [hunk 3]
```diff
@@ -134,15 +146,18 @@
...
-        // count active mobile users
-        $mobile_users = 0;
-        foreach ($ipsec_leases as $lease) {
-            if ($lease['online']) {
-                ++$mobile_users;
-            }
+    // Initialize variable unique_users
+    $unique_users = array();
+    foreach ($ipsec_leases as $lease) {
+        if ($lease['online']) {
```
Comment (AdSchellevis) [inline src/www/widgets/widgets/ipsec.widget.php:152]: I would push this one in the other loop, so you only have to travers `$ipsec_leases` once.

**Review by AdSchellevis — CHANGES_REQUESTED** [review]
Looks good, some small remarks

**Comment by Monviech** [comment]
@AdSchellevis you're always so fast, thanks! I'm happy you reviewed the code and I will improve the parts you suggested as soon as I can. If I can't figure it out I'll ask for help. Have a nice evening ^^

**Comment by AdSchellevis** [comment]
@Monviech it's a bit of a sport I guess ;) thanks for baring with me

**Comment by Monviech** [comment]
@AdSchellevis 
I have revised the code and I tested it with my fake data above, and also with real data on a production firewall. The widget still does the same job for me. Thank you for the hints to make it more efficient.

**Comment by AdSchellevis** [comment]
@Monviech looks good, thanks!

**Pull request merged**